File: adapters/bridge.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
# Copyright Contributors to the OpenTimelineIO project
"""
Bridge for gotio to access Python OTIO adapters via stdin/stdout JSON.

This script runs as a long-lived subprocess, receiving JSON-RPC style commands
on stdin and returning results on stdout. This avoids the ~100-200ms Python
startup overhead per call.

Protocol:
- One JSON object per line on stdin (request)
- One JSON object per line on stdout (response)
- Request: {"id": N, "method": "...", "params": {...}}
- Response: {"id": N, "result": ..., "error": null} or {"id": N, "result": null, "error": "..."}
"""
import json
import sys
import traceback


def main():
    """Process commands from stdin, write responses to stdout."""
    # Line-buffered output
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue

        request_id = None
        try:
            request = json.loads(line)
            request_id = request.get("id")
            response = dispatch(request)
        except Exception as e:
            response = {
                "id": request_id,
                "result": None,
                "error": f"{type(e).__name__}: {e}\n{traceback.format_exc()}",
            }

        print(json.dumps(response, separators=(",", ":")), flush=True)


def dispatch(request):
    """Dispatch a request to the appropriate handler."""
    method = request.get("method", "")
    params = request.get("params", {})
    request_id = request.get("id")

    handlers = {
        "discover": handle_discover,
        "read_from_file": handle_read_from_file,
        "read_from_string": handle_read_from_string,
        "write_to_file": handle_write_to_file,
        "write_to_string": handle_write_to_string,
        "ping": handle_ping,
    }

    handler = handlers.get(method)
    if handler is None:
        return {
            "id": request_id,
            "result": None,
            "error": f"unknown method: {method}",
        }

    try:
        result = handler(params)
        return {"id": request_id, "result": result, "error": None}
    except Exception as e:
        return {
            "id": request_id,
            "result": None,
            "error": f"{type(e).__name__}: {e}",
        }


def handle_ping(params):
    """Simple ping to check if bridge is alive."""
    return "pong"


def handle_discover(params):
    """Return list of available adapters with their capabilities."""
    import opentimelineio as otio

    result = []
    for adapter in otio.plugins.ActiveManifest().adapters:
        info = {
            "name": adapter.name,
            "suffixes": list(adapter.suffixes) if adapter.suffixes else [],
            "features": {
                "read": adapter.has_feature("read_from_file")
                or adapter.has_feature("read_from_string"),
                "write": adapter.has_feature("write_to_file")
                or adapter.has_feature("write_to_string"),
            },
        }
        result.append(info)
    return result


def handle_read_from_file(params):
    """Read a file using an adapter and return OTIO JSON."""
    import opentimelineio as otio

    filepath = params["filepath"]
    adapter_name = params.get("adapter")
    args = params.get("args", {})

    obj = otio.adapters.read_from_file(filepath, adapter_name=adapter_name, **args)
    return otio.adapters.write_to_string(obj, "otio_json")


def handle_read_from_string(params):
    """Parse a string using an adapter and return OTIO JSON."""
    import opentimelineio as otio

    data = params["data"]
    adapter_name = params["adapter"]
    args = params.get("args", {})

    obj = otio.adapters.read_from_string(data, adapter_name=adapter_name, **args)
    return otio.adapters.write_to_string(obj, "otio_json")


def handle_write_to_file(params):
    """Write OTIO JSON to a file using an adapter."""
    import opentimelineio as otio

    data = params["data"]
    filepath = params["filepath"]
    adapter_name = params.get("adapter")
    args = params.get("args", {})

    obj = otio.adapters.read_from_string(data, "otio_json")
    otio.adapters.write_to_file(obj, filepath, adapter_name=adapter_name, **args)
    return True


def handle_write_to_string(params):
    """Serialize OTIO JSON to a string using an adapter."""
    import opentimelineio as otio

    data = params["data"]
    adapter_name = params["adapter"]
    args = params.get("args", {})

    obj = otio.adapters.read_from_string(data, "otio_json")
    return otio.adapters.write_to_string(obj, adapter_name=adapter_name, **args)


if __name__ == "__main__":
    main()

File: adapters/bridge.go
// SPDX-License-Identifier: Apache-2.0
// Copyright Contributors to the OpenTimelineIO project

// Package adapters provides access to Python OTIO adapters via a bridge.
//
// This package is a compatibility layer that allows Go code to use Python
// OpenTimelineIO adapters when native Go implementations don't exist.
//
// Basic usage:
//
//	bridge, err := adapters.NewBridge(adapters.Config{})
//	if err != nil {
//		log.Fatal(err)
//	}
//	defer bridge.Close()
//
//	timeline, err := bridge.Read("project.edl")
package adapters

import (
	"bufio"
	"bytes"
	_ "embed"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"sync/atomic"

	"github.com/Avalanche-io/gotio/opentimelineio"
)

//go:embed bridge.py
var bridgeScript string

// Bridge provides access to Python OTIO adapters.
type Bridge struct {
	pythonPath string
	scriptPath string // temp file containing bridge.py
	cmd        *exec.Cmd
	stdin      io.WriteCloser
	stdout     *bufio.Reader
	stderr     bytes.Buffer
	mu         sync.Mutex // protects cmd, stdin, stdout
	requestID  atomic.Int64
	formats    []FormatInfo
	formatMap  map[string]FormatInfo // name and suffix -> info
	closed     bool
}

// Config configures the Python bridge.
type Config struct {
	PythonPath string // Path to Python interpreter (auto-detected if empty)
}

// FormatInfo describes an available file format adapter.
type FormatInfo struct {
	Name     string   // Adapter name (e.g., "otio_json", "cmx_3600")
	Suffixes []string // File extensions (e.g., [".otio", ".json"])
	CanRead  bool     // Adapter supports reading
	CanWrite bool     // Adapter supports writing
}

// request is a JSON-RPC style request to the Python bridge.
type request struct {
	ID     int64          `json:"id"`
	Method string         `json:"method"`
	Params map[string]any `json:"params"`
}

// response is a JSON-RPC style response from the Python bridge.
type response struct {
	ID     int64           `json:"id"`
	Result json.RawMessage `json:"result"`
	Error  *string         `json:"error"`
}

// Common errors.
var (
	ErrPythonNotAvailable = errors.New("Python with opentimelineio not available")
	ErrBridgeClosed       = errors.New("bridge is closed")
	ErrFormatNotFound     = errors.New("format not supported")
)

// NewBridge creates a new Python bridge.
// If cfg.PythonPath is empty, it auto-detects Python.
func NewBridge(cfg Config) (*Bridge, error) {
	pythonPath := cfg.PythonPath
	if pythonPath == "" {
		var err error
		pythonPath, err = detectPython()
		if err != nil {
			return nil, err
		}
	}

	bridge := &Bridge{
		pythonPath: pythonPath,
		formatMap:  make(map[string]FormatInfo),
	}

	if err := bridge.start(); err != nil {
		return nil, err
	}

	// Discover available formats
	if err := bridge.discover(); err != nil {
		bridge.Close()
		return nil, err
	}

	return bridge, nil
}

// detectPython finds a Python interpreter with opentimelineio installed.
func detectPython() (string, error) {
	candidates := []string{"python3", "python"}

	for _, name := range candidates {
		path, err := exec.LookPath(name)
		if err != nil {
			continue
		}

		// Verify opentimelineio is importable
		cmd := exec.Command(path, "-c", "import opentimelineio")
		if err := cmd.Run(); err == nil {
			return path, nil
		}
	}

	return "", ErrPythonNotAvailable
}

// start launches the Python subprocess.
func (b *Bridge) start() error {
	// Write bridge.py to a temp file
	tmpDir := os.TempDir()
	scriptPath := filepath.Join(tmpDir, "gotio_bridge.py")
	if err := os.WriteFile(scriptPath, []byte(bridgeScript), 0600); err != nil {
		return fmt.Errorf("failed to write bridge script: %w", err)
	}
	b.scriptPath = scriptPath

	b.cmd = exec.Command(b.pythonPath, "-u", scriptPath)
	b.cmd.Stderr = &b.stderr

	stdin, err := b.cmd.StdinPipe()
	if err != nil {
		return fmt.Errorf("failed to create stdin pipe: %w", err)
	}
	b.stdin = stdin

	stdout, err := b.cmd.StdoutPipe()
	if err != nil {
		return fmt.Errorf("failed to create stdout pipe: %w", err)
	}
	b.stdout = bufio.NewReader(stdout)

	if err := b.cmd.Start(); err != nil {
		return fmt.Errorf("failed to start Python: %w", err)
	}

	// Verify the bridge is responsive
	if err := b.ping(); err != nil {
		b.Close()
		return fmt.Errorf("bridge not responsive: %w", err)
	}

	return nil
}

// ping checks if the bridge is alive.
func (b *Bridge) ping() error {
	var result string
	if err := b.call("ping", nil, &result); err != nil {
		return err
	}
	if result != "pong" {
		return errors.New("unexpected ping response")
	}
	return nil
}

// discover fetches the list of available adapters from Python.
func (b *Bridge) discover() error {
	var adapters []struct {
		Name     string   `json:"name"`
		Suffixes []string `json:"suffixes"`
		Features struct {
			Read  bool `json:"read"`
			Write bool `json:"write"`
		} `json:"features"`
	}

	if err := b.call("discover", nil, &adapters); err != nil {
		return err
	}

	b.formats = make([]FormatInfo, 0, len(adapters))
	for _, a := range adapters {
		info := FormatInfo{
			Name:     a.Name,
			Suffixes: a.Suffixes,
			CanRead:  a.Features.Read,
			CanWrite: a.Features.Write,
		}
		b.formats = append(b.formats, info)
		b.formatMap[a.Name] = info

		// Also map by suffix
		for _, suffix := range a.Suffixes {
			s := strings.ToLower(suffix)
			if !strings.HasPrefix(s, ".") {
				s = "." + s
			}
			b.formatMap[s] = info
		}
	}

	return nil
}

// call makes a request to the Python bridge and waits for a response.
func (b *Bridge) call(method string, params map[string]any, result any) error {
	b.mu.Lock()
	defer b.mu.Unlock()

	if b.closed {
		return ErrBridgeClosed
	}

	if params == nil {
		params = make(map[string]any)
	}

	req := request{
		ID:     b.requestID.Add(1),
		Method: method,
		Params: params,
	}

	// Send request
	reqBytes, err := json.Marshal(req)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	if _, err := b.stdin.Write(append(reqBytes, '\n')); err != nil {
		return fmt.Errorf("failed to write request: %w", err)
	}

	// Read response
	line, err := b.stdout.ReadBytes('\n')
	if err != nil {
		stderr := b.stderr.String()
		if stderr != "" {
			return fmt.Errorf("bridge error: %s", stderr)
		}
		return fmt.Errorf("failed to read response: %w", err)
	}

	var resp response
	if err := json.Unmarshal(line, &resp); err != nil {
		return fmt.Errorf("failed to parse response: %w (response: %s)", err, string(line))
	}

	if resp.Error != nil {
		return errors.New(*resp.Error)
	}

	if result != nil {
		if err := json.Unmarshal(resp.Result, result); err != nil {
			return fmt.Errorf("failed to parse result: %w", err)
		}
	}

	return nil
}

// Close shuts down the Python subprocess and releases resources.
func (b *Bridge) Close() error {
	b.mu.Lock()
	defer b.mu.Unlock()

	if b.closed {
		return nil
	}
	b.closed = true

	var errs []error

	if b.stdin != nil {
		if err := b.stdin.Close(); err != nil {
			errs = append(errs, err)
		}
	}

	if b.cmd != nil && b.cmd.Process != nil {
		if err := b.cmd.Process.Kill(); err != nil {
			errs = append(errs, err)
		}
		b.cmd.Wait() // ignore error from killed process
	}

	// Clean up temp file
	if b.scriptPath != "" {
		os.Remove(b.scriptPath)
	}

	if len(errs) > 0 {
		return errs[0]
	}
	return nil
}

// AvailableFormats returns information about all available format adapters.
func (b *Bridge) AvailableFormats() []FormatInfo {
	return b.formats
}

// Read reads a file and returns an OTIO object.
// The format is auto-detected based on the file extension.
func (b *Bridge) Read(path string) (opentimelineio.SerializableObject, error) {
	ext := strings.ToLower(filepath.Ext(path))
	return b.ReadWithFormat(ext, path)
}

// Write writes an OTIO object to a file.
// The format is auto-detected based on the file extension.
func (b *Bridge) Write(obj opentimelineio.SerializableObject, path string) error {
	ext := strings.ToLower(filepath.Ext(path))
	return b.WriteWithFormat(ext, obj, path)
}

// ReadWithFormat reads a file using a specific format.
// Format can be either a suffix (e.g., ".edl") or adapter name (e.g., "cmx_3600").
func (b *Bridge) ReadWithFormat(format, path string) (opentimelineio.SerializableObject, error) {
	// Normalize format to look up in map
	f := strings.ToLower(format)
	if !strings.HasPrefix(f, ".") && b.formatMap[f].Name == "" {
		// Try adding a dot prefix
		f = "." + f
	}

	info, ok := b.formatMap[f]
	if !ok {
		return nil, fmt.Errorf("%s: %w", format, ErrFormatNotFound)
	}

	if !info.CanRead {
		return nil, fmt.Errorf("%s: format does not support reading", info.Name)
	}

	params := map[string]any{
		"filepath": path,
	}

	// If format is the adapter name, pass it explicitly
	if format == info.Name {
		params["adapter"] = info.Name
	}

	var otioJSON string
	if err := b.call("read_from_file", params, &otioJSON); err != nil {
		return nil, fmt.Errorf("python adapter error: %w", err)
	}

	return opentimelineio.FromJSONString(otioJSON)
}

// WriteWithFormat writes an OTIO object to a file using a specific format.
// Format can be either a suffix (e.g., ".edl") or adapter name (e.g., "cmx_3600").
func (b *Bridge) WriteWithFormat(format string, obj opentimelineio.SerializableObject, path string) error {
	// Normalize format to look up in map
	f := strings.ToLower(format)
	if !strings.HasPrefix(f, ".") && b.formatMap[f].Name == "" {
		// Try adding a dot prefix
		f = "." + f
	}

	info, ok := b.formatMap[f]
	if !ok {
		return fmt.Errorf("%s: %w", format, ErrFormatNotFound)
	}

	if !info.CanWrite {
		return fmt.Errorf("%s: format does not support writing", info.Name)
	}

	// Serialize to OTIO JSON
	otioJSON, err := opentimelineio.ToJSONString(obj, "")
	if err != nil {
		return fmt.Errorf("failed to serialize OTIO: %w", err)
	}

	params := map[string]any{
		"filepath": path,
		"data":     otioJSON,
	}

	// If format is the adapter name, pass it explicitly
	if format == info.Name {
		params["adapter"] = info.Name
	}

	var success bool
	if err := b.call("write_to_file", params, &success); err != nil {
		return fmt.Errorf("python adapter error: %w", err)
	}

	return nil
}
